Allow slices as well as vectors in various functions

wiktor-k opened the pull request:
I've noticed that a number of functions take owned `Vec<u8>`s but could instead take slices. This has the benefit of not requiring allocations or allowing hashing just a part of the data (subslice).

This is a WIP but if it looks good I'll adjust the documentation and also the `setkey` call.

wiktor-k committed:
Allow slices as well as vectors in `kcapi::md` module

Converted all arguments where it was possible. `KcapiHash::setkey`
unfortunately captures the `key` parameter so it needs to be a vector
for now.

This patch adds new tests for slices in doctests but leaves the old
one untouched so this change should be a non-breaking one.

Signed-off-by: Wiktor Kwapisiewicz <wiktor@metacode.biz>

puru1761 commented:
There are other places where we use vectors in place of [u8] slices. Can you change those as well? Since these are breaking changes, we may have to update the crate as well when you are done.

wiktor-k commented:
I had yet another idea that'd allow using slices but would still accept vectors (not to introduce a breaking change): using `impl AsRef<[u8]>` instead of Vec.

Attached code so you can see how it works (as well as extended code example to include u8 slice).

I'm OK with either approach (convert to u8 slices or impl AsRef) but this reduces the number of breaking changes.

Thanks for help! :wave:

puru1761 commented:
Hi Wiktor,

Could you please confirm if the impl Asref<[u8]> is an approach that does not break the existing implementation?
I would like to not have to update my downstream consumer of this crate, so it would be best if you could add a few tests
to demonstrate that.

~ Puru

wiktor-k commented:
> so it would be best if you could add a few tests to demonstrate that.

Actually this already is covered by doctests (the old tests that use vectors didn't need to be adjusted). I'll convert all other use-sites to that...

wiktor-k committed:
Allow slices as well as vectors in `kcapi::skcipher` module

Converted all arguments where it was possible. `IOVec` parameters left
as they are.

This patch adds new tests for slices in doctests but leaves the old
one untouched so this change should be a non-breaking one.

Signed-off-by: Wiktor Kwapisiewicz <wiktor@metacode.biz>

wiktor-k committed:
Allow slices as well as vectors in `kcapi::kdf` module

Converted all arguments where it was possible.

This patch adds new tests for slices in doctests but leaves the old
one untouched so this change should be a non-breaking one.

Signed-off-by: Wiktor Kwapisiewicz <wiktor@metacode.biz>

wiktor-k committed:
Allow slices as well as vectors in `kcapi::aead::ccm_nonce_to_iv`

This patch adds new tests for slices in doctests but leaves the old
one untouched so this change should be a non-breaking one.

Signed-off-by: Wiktor Kwapisiewicz <wiktor@metacode.biz>

wiktor-k committed:
Allow slices as well as vectors in `kcapi::akcipher` module

Converted all arguments where it was possible. Arguments that are
persisted as part of the structures were left as `Vec<u8>`s.

Signed-off-by: Wiktor Kwapisiewicz <wiktor@metacode.biz>

wiktor-k commented:
Okay, it should be all low-hanging fruits.

If some structs did not capture `key`s and other things then it could be optimized further but for this PR I wanted to preserve the API.

It should be ready for review. Thanks! :pray:
